common/wuanalysis/anawu.cpp

```diff
@@ -0,0 +1,429 @@
+}
+
+void WuScope::setOutput(unsigned i, WuScope * scope)
+{
+    while (outputs.size() <= i)
+        outputs.push_back(nullptr);
+    outputs[i] = scope;
+}
+
+WuScope * WuScope::querySource()
+{
+    const char * source = attrs->queryProp("@IdSource");
+    if (!source)
+        return nullptr;
+    return parent->select(source);
+}
+
+WuScope * WuScope::queryTarget()
+{
+    const char * target = attrs->queryProp("@IdTarget");
+    if (!target)
+        return nullptr;
+    return parent->select(target);
+}
+
+stat_t WuScope::getStatRaw(StatisticKind kind, StatisticKind variant) const
+{
+    StringBuffer name;
+    name.append('@').append(queryStatisticName(kind | variant));
+    return attrs->getPropInt64(name);
+}
+
+unsigned WuScope::getAttr(WuAttr attr) const
+{
+    StringBuffer name;
+    name.append('@').append(queryWuAttributeName(attr));
+    return attrs->getPropInt64(name);
+}
+
+void WuScope::getAttr(StringBuffer & result, WuAttr attr) const
+{
+    StringBuffer name;
+    name.append('@').append(queryWuAttributeName(attr));
+    attrs->getProp(result, name);
+}
+
+void WuScope::trace(unsigned indent) const
+{
+    printf("%*s%s: \"%s\" (", indent, " ", queryScopeTypeName(), queryName());
+    for (auto in : inputs)
+        printf("%s ", in->queryName());
+    printf("->");
+    for (auto out : outputs)
+        printf(" %s",  out->queryName());
+    printf(") [");
+    printf("children: ");
+    for(auto & scope: scopes)
+    {
+        printf("%s ", scope.queryName());
+    }
+    printf("] {\n");
+
+    printXML(queryAttrs(), indent);
+    for (auto & cur : scopes)
+    {
+        cur.trace(indent+4);
+    }
+    printf("%*s}\n",indent, " ");
+}
+
+
+//-----------------------------------------------------------------------------------------------------------
+/* Callback used to output scope properties as xml */
+class StatsGatherer : public IWuScopeVisitor
+{
+public:
+    StatsGatherer(IPropertyTree * _scope) : scope(_scope){}
+
+    virtual void noteStatistic(StatisticKind kind, unsigned __int64 value, IConstWUStatistic & cur) override
+    {
+        StringBuffer name;
+        name.append('@').append(queryStatisticName(kind));
+        scope->setPropInt64(name, value);
+    }
+    virtual void noteAttribute(WuAttr attr, const char * value)
+    {
+        StringBuffer name;
+        name.append('@').append(queryWuAttributeName(attr));
+        scope->setProp(name, value);
+    }
+    virtual void noteHint(const char * kind, const char * value)
+    {
+        throwUnexpected();
+    }
+    IPropertyTree * scope;
+};
+
+
+//-----------------------------------------------------------------------------------------------------------
+WorkunitAnalyser::WorkunitAnalyser(WuAnalyseOptions & _options) : root("", nullptr), options(_options)
+{
+    gatherRules(rules);
+}
+
+void WorkunitAnalyser::check(const char * scope, IWuActivity & activity)
+{
+    if (activity.queryThorActivityKind()!=TAKhashdistribute)
```

Comment:
This test is a bit arbitrary - it suggests the minimum time test should be done somewhere else, or in a different way.

Reply:
The time taken to distribute does not seem relevant to skewed distribution rows.  We could check the minimum time anyway for hash distributes.  What do you think?